neurobooth_os/tasks/DSC.py

```diff
@@ -123,6 +123,17 @@ def run(self, prompt=True, last_task=False, subj_id="test", **kwargs):
         self.present_complete(last_task)
         return self.events
 
+    def repeat_advance(self):
```

Comment:
I'd argue that we should not alter/repeat the advance/repeat logic in subclasses. Instead, I think it makes sense to let the repeat/advance logic be in the task base class and define the key sets as class attributes that can be overridden by the subclass.

Ex:
```
def repeat_advance(self) -> bool:
        """
         Repeat the current task or continue to next, based on the key pressed.
         :returns: False to continue; True to repeat
         """
        keys = utils.get_keys(keyList=[*self.advance_keys, *self.repeat_keys])
        for key in keys:
            if key in self.advance_keys:
                return False
            elif key in self.repeat_keys:
                return True
        self.logger.error('Unreachable case during task repeat_advance: keys={keys}')
```

Then in `__init__`:
```
self.advance_keys = ['space']
self.repeat_keys = ['r', ',']
```

In the subclass `__init__` after calling `super`:
```
self.repeat_keys = ['r']
```

Reply:
Of course, would need to alter slightly to take into account the new keyword argument. Could toggle in Task base class based on that argument instead of override in subclass.

Reply:
I think toggling behavior in a superclass to handle differences between subtypes is an anti-pattern, as it subverts the class hierarchy. 

In general, having to look at the DB to determine the correct behavior for a task seems somewhere between an anti-pattern and a necessary evil. 

I will add the log statement: 
>         self.logger.error('Unreachable case during task repeat_advance: keys={keys}')
But I'm not sure this rises to the level of an error, since hitting some other key at this phase seems like it could be a fairly common occurrence.  I'll make it a warning, unless I'm missing something.

Reply:
1. My main concern is the re-implementation of repeat_advance logic in the subclasses. I think the following in the base class constructor would address both our concerns:
```
self.advance_keys = ['space']
if task_repeatable_by_subject:
    self.repeat_keys = ['r', ',']
else:
    self.repeat_keys = ['r']
```

2. I agree that needing to look at the DB as somewhat evil, but that is currently the paradigm for handling task configuration. (And I consider `task_repeatable_by_subject` to be a generic task customization option that we are just happening to apply to MOT and DSC. There could be other tasks in the future where this is desirable from an experimental perspective.) Hopefully DB lookups will all be replaced by some sort of config file.

3. It should truly be an unreachable line error in the for-loop formulation I suggested above, as we would have to exhaust all keys simultaneously pressed before it triggered. (And my understandings is that get_keys should only return if at least one of the keys is in the provided list.)